IM/connectors/OCCI.py

```diff
@@ -1271,6 +1275,57 @@ def get_keystone_token(occi, keystone_uri, auth):
         """
         Contact the specified keystone server to return the token
         """
+        version = KeyStoneAuth.get_keystone_version(occi, keystone_uri, auth)
+        if version == 2:
+            occi.logger.debug("Getting Keystone v2 token")
+            return KeyStoneAuth.get_keystone_token_v2(occi, keystone_uri, auth)
+        elif version == 3:
+            occi.logger.debug("Getting Keystone v3 token")
+            return KeyStoneAuth.get_keystone_token_v3(occi, keystone_uri, auth)
+        else:
+            # this must never happen
+            raise Exception("Error obtaining Keystone Token: Unknown version %d" % version)
+
+    @staticmethod
+    def get_keystone_version(occi, keystone_uri, auth):
+        """
+        Contact the specified keystone server to return version to use
+        """
+        version = None
+        token = auth and "token" in auth
+
+        try:
+            uri = uriparse(keystone_uri)
+            server = uri[1].split(":")[0]
+            port = int(uri[1].split(":")[1])
+
+            url = "https://%s:%s" % (server, port)
+            resp = occi.create_request_static('GET', url, None, {})
+            if resp.status_code == 200:
+                json_data = resp.json()
+            for elem in json_data["versions"]["values"]:
```

Comment:
If the Keystone URI is already including the /v3 or /v2.0 this will probably cause an exception

Reply:
I would consider this a site configuration mistake ... isn't there an official recommendation against configuring Keystone redirect URI with a version already included?

Reply:
So yes, it's a configuration mistake, but see https://docs.openstack.org/keystonemiddleware/latest/middlewarearchitecture.html#configuration

Extracted from there:
# Complete "public" Identity API endpoint. This endpoint should not be an
# "admin" endpoint, as it should be accessible by all end users.
# Unauthenticated clients are redirected to this endpoint to authenticate.
# Although this endpoint should ideally be unversioned, client support in the
# wild varies. If you're using a versioned v2 endpoint here, then this should
# *not* be the same endpoint the service user utilizes for validating tokens,
# because normal end users may not be able to reach that endpoint. (string
# value)

Let's assume we do not have faulty clients so the code above is correct :)